pkg/signature/algorithm_registry.go

```diff
@@ -0,0 +1,259 @@
+	}
+	return rsaKeySize, nil
+}
+
+func (a algorithmDetailsImpl) GetECDSACurve() (*elliptic.Curve, error) {
+	if a.keyType != ECDSA {
+		return nil, fmt.Errorf("unable to retrieve ECDSA curve for key type: %T", a.keyType)
+	}
+	ecdsaCurve, ok := a.extraKeyParams.(elliptic.Curve)
+	if !ok {
+		// This should be unreachable.
+		return nil, fmt.Errorf("unable to retrieve curve for ECDSA, malformed algorithm details?: %T", a.keyType)
+	}
+	return &ecdsaCurve, nil
+}
+
+func (a algorithmDetailsImpl) checkKey(pubKey crypto.PublicKey) (bool, error) {
+	switch a.keyType {
+	case RSA:
+		rsaKey, ok := pubKey.(*rsa.PublicKey)
+		if !ok {
+			return false, nil
+		}
+		keySize, err := a.GetRSAKeySize()
+		if err != nil {
+			return false, err
+		}
+		return rsaKey.Size() == int(keySize), nil
+	case ECDSA:
+		ecdsaKey, ok := pubKey.(*ecdsa.PublicKey)
+		if !ok {
+			return false, nil
+		}
+		curve, err := a.GetECDSACurve()
+		if err != nil {
+			return false, err
+		}
+		return ecdsaKey.Curve == *curve, nil
+	case ED25519:
+		_, ok := pubKey.(ed25519.PublicKey)
+		return ok, nil
+	}
+	return false, fmt.Errorf("unrecognized key type: %T", a.keyType)
+}
+
+func (a algorithmDetailsImpl) checkHash(hashType crypto.Hash) bool {
+	return a.hashType == hashType
+}
+
+var algorithmDetails = []algorithmDetailsImpl{
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_2048_SHA256, RSA, crypto.SHA256, RSAKeySize(2048), "rsa-sign-pkcs1-2048-sha256"},
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_3072_SHA256, RSA, crypto.SHA256, RSAKeySize(3072), "rsa-sign-pkcs1-3072-sha256"},
+	{v1.KnownSignatureAlgorithm_RSA_SIGN_PKCS1_4096_SHA256, RSA, crypto.SHA256, RSAKeySize(4096), "rsa-sign-pkcs1-4096-sha256"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_256_NISTP256, ECDSA, crypto.SHA256, elliptic.P256(), "ecdsa-sha2-256-nistp256"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_384_NISTP384, ECDSA, crypto.SHA384, elliptic.P384(), "ecdsa-sha2-384-nistp384"},
+	{v1.KnownSignatureAlgorithm_ECDSA_SHA2_512_NISTP521, ECDSA, crypto.SHA512, elliptic.P521(), "ecdsa-sha2-512-nistp521"},
+	{v1.KnownSignatureAlgorithm_ED25519, ED25519, crypto.Hash(0), nil, "ed25519"},
+	{v1.KnownSignatureAlgorithm_ED25519_PH, ED25519, crypto.SHA512, nil, "ed25519-ph"},
+}
+
+// AlgorithmRegistryConfig represents a set of permitted algorithms for a given Sigstore service or component.
+//
+// Individual services may wish to restrict what algorithms are allowed to a subset of what is covered in the algorithm
+// registry (represented by v1.KnownSignatureAlgorithm).
+type AlgorithmRegistryConfig struct {
+	permittedAlgorithms []AlgorithmDetails
+}
+
+// GetAlgorithmDetails retrieves a set of details for a given v1.KnownSignatureAlgorithm flag that allows users to
+// introspect the public key algorithm, hash algorithm and more.
+func GetAlgorithmDetails(knownSignatureAlgorithm v1.KnownSignatureAlgorithm) (AlgorithmDetails, error) {
+	for _, detail := range algorithmDetails {
```

Comment:
maybe turn this into a map[v1.KnownSignatureAlgorithm]AlgorithmDetails ?

Reply:
I was thinking about this, but my reasons not to were:
- If I do that, I have to match the `v1.KnownSignatureAlgorithm` in the `AlgorithmDetails` with the one in the map key, which is another opportunity for subtle typos that can introduce bugs. I can't remove it from `AlgorithmDetails` since we need it for things like `AlgorithmRegistryConfig.ListPermittedAlgorithms`.
- The `algorithmDetails` array only includes an entry for each signature algorithm in the registry which by design, is a small, constrained set of algorithms. I haven't benchmarked and to be honest, I don't know much about Go's performance characteristics, but I wouldn't expect the map to be faster than a simple linear search for this number of elements.
- This array isn't part of the API so if my assumptions are wrong, we can easily walk this decision back if we need to without introducing breaking changes.

Happy to change it over to a map, if you'd prefer though.